### PR TITLE
[chore] Implement entity dtype validation to enforce dtype consistency

### PR DESCRIPTION
**Pull request by gee-senbong**
## Description

This PR implements entity dtype validation. When entity is associated with a table column,
* it will initiate entity dtype if the entity has been used by another table column;
* it will throw error if entity dtype inconsistency is detected (user has to resolve the inconsistency first before using it in any other table column);
* checks whether the dtype of the column to be associated is consistent with existing column(s) of the same entity.

## Type of Change

Label this pull request to place it under the correct category in Release Notes:

|        **Pull Request Label**         | **Category in Release Notes** |
|:-------------------------------------:|:-----------------------------:|
|       `enhancement`, `feature`        |          🚀 Features          |
| `bug`, `refactoring`, `bugfix`, `fix` |    🔧 Fixes & Refactoring     |
|       `build`, `ci`, `testing`        |    📦 Build System & CI/CD    |
|              `breaking`               |      💥 Breaking Changes      |
|            `documentation`            |       📝 Documentation        |
|            `dependencies`             |    ⬆️ Dependencies updates    |

## Checklist

- [ ] I have read the [`CODE_OF_CONDUCT.md`](https://github.com/featurebyte/featurebyte/blob/main/CODE_OF_CONDUCT.md) and [`CONTRIBUTING.md`](https://github.com/featurebyte/featurebyte/blob/main/CONTRIBUTING.md) guides.
- [ ] I have written tests for the changes made.
- [ ] I have written docstrings in [NumpyDoc format](https://numpydoc.readthedocs.io/en/latest/format.html#docstring-standard)
- [ ] I have labeled my Pull Request correctly

### CODE REVIEW

File: featurebyte/models/entity.py [hunk 1]
```diff
@@ -83,6 +83,7 @@ class EntityModel(EntityRelationship):
         ID of table with primary key columns associated to the entity
     """
 
+    dtype: Optional[DBVarType] = None
```
Comment (YS-L) [inline featurebyte/models/entity.py:86]: Curious what is the difference between declaring this as None directly vs `Field(default=None)`?

File: featurebyte/service/table_columns_info.py [hunk 2]
```diff
@@ -37,6 +37,146 @@
...
+                    await self.entity_service.update_document(
+                        document_id=col_info.entity_id,
+                        data=EntityServiceUpdate(dtype=col_info.dtype),
+                        return_document=False,
+                    )
+
+            if prev_entity_id and col_info.entity_id is None:
+                # when the entity is disassociated from a column, update the entity dtype if needed
+                entity = await self.entity_service.get_document(prev_entity_id)
+                if not entity.table_ids:
```
Comment (YS-L) [inline featurebyte/service/table_columns_info.py:173]: The `table_ids` of the entity are updated first before this part, is that right?

File: tests/unit/service/test_table_columns_info.py [hunk 3]
```diff
@@ -644,3 +645,231 @@ async def test_scd_table_update_columns_info__entity_relationship(
...
+    with pytest.raises(DocumentUpdateError) as exc:
+        await app_container.table_columns_info_service.update_columns_info(
+            service=app_container.dimension_table_service,
+            document_id=dimension_table.id,
+            columns_info=construct_columns_info(dimension_table, "col_int", entity_foo.id),
+        )
+
+    expected_msg = (
+        f"Entity foo (ID: {entity_foo.id}) has columns with different dtypes (VARCHAR and INT) in "
+        f"tables sf_scd_table and sf_event_table. Please double-check the entity of the affected columns."
```
Comment (YS-L) [inline tests/unit/service/test_table_columns_info.py:849]: nit: would associating the dtypes with the table names be clearer? e.g.

> Entity foo (ID: {entity_foo.id}) has columns with different dtypes in tables sf_scd_table (VARCHAR) and sf_event_table (INT). Please double-check the entity of the affected columns.

**Review by YS-L — APPROVED** [review]

File: featurebyte/models/entity.py (hunk 1 above)
Comment (gee-senbong) [reply featurebyte/models/entity.py:86]: `Field(default=None)` is for the old pydantic version, using `None` is supported & suggested for the new pydantic.

File: featurebyte/service/table_columns_info.py (hunk 2 above)
Comment (gee-senbong) [reply featurebyte/service/table_columns_info.py:173]: Yes, `table_ids` is updated before calling this. I should update the doc string here to make it clearer.

**Pull request merged**